students/jbutts/Lesson09/mailroom_oo.py

```diff
@@ -0,0 +1,149 @@
+                    # alphabetical by keys. Write out rows: "Donor", "Total", "Gifts", "Average"
+                    row = key, sum(value), len(value), sum(value)/len(value)
+                    donors.append("{:<25s}${:<19.2f}{:<10d}${:<13.2f}\n".format(*row))
+        return donors
+
+    def show_donors(self):
+        '''
+        Write a report to screen of donors and information about how they've donated
+
+        Return values in descending order from the sum of their historical contributions
+        '''
+
+        # Write header row
+        row = "Donor", "Total", "Gifts", "Average"
+        print("{:<25s}{:<20s}{:<10s}{:<13s}".format(*row))
+        # Write corresponding ===='s
+        print(''.join(("=" * 24 + " ",  # donor
+                       "=" * 19 + " ",  # total
+                       "=" * 9 + " ",   # gifts
+                       "=" * 12)))      # average
+
+        print(''.join(self.get_donors_details()))
+
+    def list_donors(self):
+            '''
+            Write a numbered list for the purpose of selecting a donor
+
+            Use a list comprehension to print the numbered list
+            '''
+            # List current donors in the dictionary as '#. First Last'
+            donor_list = "{:<4} {:<7}\n".format("No.", "Name")
+            donor_list += "{} {}\n".format("=" * 4, "=" * 7)
+
+            for i, donor in enumerate(sorted(self.donor_db)):
+                donor_list += "{:<4} {:<7}\n".format(i + 1, donor)
+
+            return donor_list
+
+    def add_contribution(self, name, amount):
+        '''
+        After record_contribution() prompts user for input, add records to dictionary as needed.
+
+        Do extra handling to account for floats and lists for the amounts
+        '''
+
+        if self.donor_db.get(name):
+            if isinstance(self.donor_db[name], list):
+                if isinstance(amount, list):
+                    self.donor_db[name].extend(amount)  # when amount is a list
+                else:
+                    self.donor_db[name].extend([amount])  # in case we get passed a float for amount
+        else:
+            # otherwise, add a single contribution with a new user
+            if isinstance(amount, list):
+                self.donor_db[name] = amount
+            else:
+                self.donor_db[name] = [amount]
+        return self.donor_db
+
+    def send_letter_to_all(self):
+        #  Write text files containing letter contents for all donors in the database
+
+
+        for key, value in self.donor_db.items():
+            tempfilepath = tempfile.gettempdir() + "/" + str(key.replace(' ', "_")) + ".txt"
+            try:
+                with open(tempfilepath, "w") as outputfile:
+                    outputfile.writelines(Donor.format_email(self, key, value[len(value)-1], sum(self.donor_db[key])))
+            except IOError:
+                print("IOError: Could not create file:" + tempfilepath)
+
+
+class Donor(object):
```

Comment:
Donor class isn't being properly utilized.

See example here:
https://repl.it/@naaleksandrova/Lesson9Example